packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+        require(
+            redemptions[redemptionId].expiryTime != 0 && getCurrentTime() >= redemptions[redemptionId].expiryTime,
+            "unexpired or nonexistent"
+        );
+
+        for (uint256 i = 0; i < cumulativeRedemptions.length; i++) {
+            IERC20 token = cumulativeRedemptions[i].token;
+            uint256 currentRedemptionTotal = redeemedAmounts[tokenId][token];
+            uint256 proposedRedemptionTotal = cumulativeRedemptions[i].amount;
+
+            // Only pay if the cumulative amount specified in the request is larger than the amount paid out already.
+            if (proposedRedemptionTotal > currentRedemptionTotal) {
+                uint256 amountToPay = proposedRedemptionTotal - currentRedemptionTotal;
+                redeemedAmounts[tokenId][token] = proposedRedemptionTotal;
+                token.safeTransfer(msg.sender, amountToPay);
+            }
+        }
+
+        // Return the bond to the owner.
+        bondToken.safeTransfer(msg.sender, redemptions[redemptionId].expiryTime + bond);
+
+        delete redemptions[redemptionId];
+
+        emit Redeemed(tokenId, redemptionId, redemptions[redemptionId].expiryTime);
+    }
+
+    /**
+     * @notice gets the current time. Can be overridden for testing.
+     */
+    function getCurrentTime() public view virtual returns (uint256) {
+        return block.timestamp;
+    }
+
+    /**
+     * @notice Abstract function that is called to mint the next ERC721 tokenId.
+     * @param recipient the recipient of the newly minted token.
+     */
+    function mintNextToken(address recipient) public virtual returns (uint256);
+
+    /**
+     * @notice Abstract function that is called to check the owner of the token.
+     * @dev this matches the ERC721 ownerOf interface.
+     * @param tokenId the tokenId to check the owner of.
+     */
+    function ownerOf(uint256 tokenId) public view virtual returns (address);
+
+    function _redemptionId(uint256 tokenId, RedemptionAmount[] memory amounts) internal pure returns (bytes32) {
+        return keccak256(abi.encode(tokenId, amounts));
+    }
+
+    function _getStore() internal view returns (StoreInterface) {
+        return StoreInterface(finder.getImplementationAddress(OracleInterfaces.Store));
+    }
+
+    function _getOptimisticOracle() internal view returns (SkinnyOptimisticOracleInterface) {
+        return
+            SkinnyOptimisticOracleInterface(finder.getImplementationAddress(OracleInterfaces.SkinnyOptimisticOracle));
+    }
+
+    function _getIdentifierWhitelist() internal view returns (IdentifierWhitelistInterface) {
+        return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
+    }
+
+    function _getCollateralWhitelist() internal view returns (AddressWhitelistInterface) {
+        return AddressWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.CollateralWhitelist));
+    }
```

Comment:
these methods introduce a fair amount of gas overhead. do we want to consider the cache and sync pattern that the across contracts use?

gas might be less of a consideration here but it will still be pretty meaningful!

Reply:
Makes sense. We can definitely do this.

Reply:
Note: some of these are only used in the constructor, so those don't need to be cached.

Reply:
Done.